src/main/java/net/minecraftforge/client/event/RenderLevelStageEvent.java

```diff
@@ -0,0 +1,145 @@
+package net.minecraftforge.client.event;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import javax.annotation.Nullable;
+
+import com.mojang.blaze3d.vertex.PoseStack;
+import com.mojang.math.Matrix4f;
+
+import net.minecraft.client.renderer.LevelRenderer;
+import net.minecraft.client.renderer.RenderType;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraftforge.eventbus.api.Event;
+import net.minecraftforge.fml.event.IModBusEvent;
+
+/**
+ * Fires at various times during LevelRenderer.renderLevel. 
+ * Check {@link #stage} to render during the appropriate time for your use case.
+ */
+public class RenderLevelStageEvent extends Event
+{
+    public final Stage stage;
+    public final LevelRenderer levelRenderer;
+    public final PoseStack poseStack;
+    public final Matrix4f projectionMatrix;
+    public final int ticks;
+    public final float partialTick;
+    public final double camX, camY, camZ;
+
+    public RenderLevelStageEvent(Stage stage, LevelRenderer levelRenderer, PoseStack poseStack, Matrix4f projectionMatrix, int ticks, float partialTick, double camX, double camY, double camZ)
+    {
+        this.stage = stage;
+        this.levelRenderer = levelRenderer;
+        this.poseStack = poseStack;
+        this.projectionMatrix = projectionMatrix;
+        this.ticks = ticks;
+        this.partialTick = partialTick;
+        this.camX = camX;
+        this.camY = camY;
+        this.camZ = camZ;
+    }
+
+    /**
+     * Use to create a custom {@link RenderLevelStageEvent.Stage}s.
+     * Fired after the LevelRenderer has been created.
+     */
+    public static class RegisterStageEvent extends Event implements IModBusEvent
```

Comment:
I'm a bit unsure on putting a mod-bus event inside a non-mod-bus event, because of the potential confusion by modders. I won't say this event should be moved, but at least it should be noted in the javadoc.